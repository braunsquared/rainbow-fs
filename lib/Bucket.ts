import fs from 'fs-extra';
import yaml from 'js-yaml';
import fsPath from 'path';
import glob from 'glob';
import util from 'util';

import { IStore } from './Store';
import { Path } from './impl/sitecore/Path';
import { VirtualItem } from './impl/sitecore/VirtualItem';
import { HintFilename, IItem, IPath } from './model/Sitecore';

export type ErrorCallbackValue = Error | null | undefined;

const readFile = util.promisify<string, Buffer>(fs.readFile);
const writeFile = util.promisify<string, string>(fs.writeFile);
const mkdirp = util.promisify<string>(fs.mkdirp);
const asyncGlob = util.promisify<string, string[]>(glob);

export class Bucket{
  Filepath: string;
  Path: Path;
  DB: string;

  constructor(filepath: string, sitecorePath: string, db: string) {
    this.Filepath = filepath;
    this.Path = new Path(sitecorePath);
    this.DB = db;
  }

  convertItemPath(scPath: IPath): string {
    const deltaPath = new Path(scPath.Folder.substring(this.Path.Path.length));
    return fsPath.join(this.Filepath, deltaPath.Path, `${scPath.Name}.yml`);
  }

  async read(store: IStore) {
    if (!this.Filepath) {
      throw new Error("Filepath is not set. Unable to read from file system.");
    }

    const files = await asyncGlob(fsPath.join(this.Filepath, '**/*.yml'));

    for(let i = 0; i < files.length; ++i) {
      const file = files[i];
      const buffer: Buffer = await readFile(file);
      const content = yaml.safeLoad(buffer.toString('utf8'), {filename: file});
      const item = store.createItemFromObject(content, {
        isNew: false,
        isDirty: false,
        hints: new Map([[HintFilename, file]]),
      });

      if(item.DB !== this.DB) {
        console.warn('Found item in bucket from different DB');
        return;
      }

      if(item.Path.Folder === this.Path.Path && item.Parent) {
        // We may be able to infer some information about our item
        const db = store.getDatabase(item.DB);
        if(!db) {
          throw new Error('Database should not be null');
        }

        let bucketItem = db.getItem(item.Parent);
        if (!bucketItem) {
          bucketItem = new VirtualItem(item.Parent, this.Path.Path);
          db.addItem(bucketItem);
        }
      }
    }
  }

  async write(item: IItem) {
    if(item.Meta.isVirtual) {
      throw new Error('Unable to serialize virtual items');
    } else if(!item.Path.Path.startsWith(this.Path.Path)) {
      throw new Error('Item is not a descendant of this bucket');
    } else if(item.DB !== this.DB) {
      throw new Error('Item DB does not match the Bucket DB');
    }

    // Figure out what the path should be
    const itemPath = this.convertItemPath(item.Path);
    const basePath = fsPath.dirname(itemPath);

    await mkdirp(basePath);
    await writeFile(itemPath, yaml.safeDump(item.toObject()));
  }
}
