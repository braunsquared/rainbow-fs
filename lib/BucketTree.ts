import { Bucket } from './Bucket';

export type BucketWalkCallback = (bucket: Bucket) => void;

export class BucketEntry {
  readonly path: string;
  readonly parent: BucketEntry | null;
  readonly bucket: Bucket | null;
  readonly children: Array<BucketEntry>;

  constructor(path: string, parent: BucketEntry | null = null, bucket: Bucket | null = null) {
    this.path = path;
    this.parent = parent;
    this.bucket = bucket;
    this.children = [];
  }

  bucketForFSPath(fsPath: string): Bucket | undefined {
    if (this.bucket && this.bucket.Filepath === fsPath) {
      return this.bucket;
    }

    let childBucket;

    this.children.find(c => {
      childBucket = c.bucketForFSPath(fsPath);
      return !!childBucket;
    });

    return childBucket;
  }

  addBucket(path: string, bucket: Bucket): void {
    if(!path) {
      throw new Error('Invalid path. Duplicate bucket path or you failed to pass one');
    }

    const closest = this.closest(path);
    if (closest) {
      return closest.addBucket(path.substring(closest.path.length), bucket);
    }

    // We don't have a closest, so lets figure out where we are remapping our children
    const entry = new BucketEntry(path, this, bucket);
    this.children
      .filter(child => child.path.indexOf(path) === 0)
      .forEach(child => {
        entry.addBucket(child.path.substring(path.length), child.bucket as Bucket)
        this.children.splice(this.children.indexOf(child), 1);
      });
    
    this.children.push(entry);
  }

  closest(path: string): BucketEntry | null {
    const entry = this.children.find(child => path.indexOf(child.path) === 0);
    if (!entry) {
      return null;
    }

    return entry.closest(path.substring(entry.path.length + 1)) || entry;
  }

  walk(cb: BucketWalkCallback) {
    if (this.bucket) {
      cb(this.bucket);
    }

    this.children.forEach(entry => {
      entry.walk(cb);
    });
  }
}

export class BucketTree {
  readonly root: BucketEntry;

  constructor() {
    this.root = new BucketEntry('');
  }

  addBucket(bucket: Bucket) {
    const bucketPath = bucket.Path;
    if (!bucketPath) {
      throw new Error('Bucket must have a path.');
    }

    const existingBucket = this.root.bucketForFSPath(bucket.Filepath);
    if(existingBucket) {
      throw new Error('Duplicate file system path');
    }

    this.root.addBucket(bucketPath.Path, bucket);
  }

  closest(path: string): Bucket | null {
    const closestItem = this.root.closest(path);
    return closestItem ? closestItem.bucket : null;
  }

  walk(cb: BucketWalkCallback) {
    this.root.walk(cb);
  }
}
