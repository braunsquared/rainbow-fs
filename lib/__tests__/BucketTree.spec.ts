import { BucketTree } from "../BucketTree";
import { Bucket } from "../Bucket";

describe('BucketTree', () => {
  let tree: BucketTree;

  beforeEach(() => {
    tree = new BucketTree();
  });

  it('should contain a single bucket', () => {
    tree.addBucket(new Bucket('/data/content', '/sitecore/content', 'master'));
    expect(tree).toMatchSnapshot();
  });

  it('should contain 2 buckets which are siblings', () => {
    tree.addBucket(new Bucket('/data/content', '/sitecore/content/', 'master'));
    tree.addBucket(new Bucket('/data/templates', '/sitecore/templates/', 'master'));
    expect(tree).toMatchSnapshot();
  });

  it('should have a tree of buckets with descendants', () => {
    tree.addBucket(new Bucket('/data/content', '/sitecore/content/', 'master'));
    tree.addBucket(new Bucket('/data/templates', '/sitecore/templates/', 'master'));
    tree.addBucket(new Bucket('/data/helix/content', '/sitecore/content/Helix/', 'master'));
    tree.addBucket(new Bucket('/data/helix/project/test', '/sitecore/content/Helix/Project/TestProject/', 'master'));
    tree.addBucket(new Bucket('/data/helix/foundation/layer.project', '/sitecore/content/Helix/Project/', 'master'));
    tree.addBucket(new Bucket('/data/helix/foundation/layer.foundation', '/sitecore/content/Helix/Foundation/', 'master'));
    expect(tree).toMatchSnapshot();
  });

  it('should fail to add a bucket with a duplicate sitecore path', () => {
    expect(() => {
      tree.addBucket(new Bucket('/data/content1', '/sitecore/content/', 'master'));
      tree.addBucket(new Bucket('/data/content2', '/sitecore/content/', 'master'));
    }).toThrow();
  });

  it('should fail to add a bucket with a duplicate file system path', () => {
    expect(() => {
      tree.addBucket(new Bucket('/data/content1', '/sitecore/content/bucket1', 'master'));
      tree.addBucket(new Bucket('/data/content1', '/sitecore/content/bucket2', 'master'));
    }).toThrow();
  });
});
