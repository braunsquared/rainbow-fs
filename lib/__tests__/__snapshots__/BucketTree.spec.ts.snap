// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BucketTree should contain 2 buckets which are siblings 1`] = `
BucketTree {
  "root": BucketEntry {
    "bucket": null,
    "children": Array [
      BucketEntry {
        "bucket": Bucket {
          "DB": "master",
          "Filepath": "/data/content",
          "Path": Path {
            "Folder": "/sitecore",
            "Name": "content",
            "Path": "/sitecore/content",
            "Tokens": Array [
              "sitecore",
              "content",
            ],
          },
        },
        "children": Array [],
        "parent": [Circular],
        "path": "/sitecore/content",
      },
      BucketEntry {
        "bucket": Bucket {
          "DB": "master",
          "Filepath": "/data/templates",
          "Path": Path {
            "Folder": "/sitecore",
            "Name": "templates",
            "Path": "/sitecore/templates",
            "Tokens": Array [
              "sitecore",
              "templates",
            ],
          },
        },
        "children": Array [],
        "parent": [Circular],
        "path": "/sitecore/templates",
      },
    ],
    "parent": null,
    "path": "",
  },
}
`;

exports[`BucketTree should contain a single bucket 1`] = `
BucketTree {
  "root": BucketEntry {
    "bucket": null,
    "children": Array [
      BucketEntry {
        "bucket": Bucket {
          "DB": "master",
          "Filepath": "/data/content",
          "Path": Path {
            "Folder": "/sitecore",
            "Name": "content",
            "Path": "/sitecore/content",
            "Tokens": Array [
              "sitecore",
              "content",
            ],
          },
        },
        "children": Array [],
        "parent": [Circular],
        "path": "/sitecore/content",
      },
    ],
    "parent": null,
    "path": "",
  },
}
`;

exports[`BucketTree should have a tree of buckets with descendants 1`] = `
BucketTree {
  "root": BucketEntry {
    "bucket": null,
    "children": Array [
      BucketEntry {
        "bucket": Bucket {
          "DB": "master",
          "Filepath": "/data/content",
          "Path": Path {
            "Folder": "/sitecore",
            "Name": "content",
            "Path": "/sitecore/content",
            "Tokens": Array [
              "sitecore",
              "content",
            ],
          },
        },
        "children": Array [
          BucketEntry {
            "bucket": Bucket {
              "DB": "master",
              "Filepath": "/data/helix/content",
              "Path": Path {
                "Folder": "/sitecore/content",
                "Name": "Helix",
                "Path": "/sitecore/content/Helix",
                "Tokens": Array [
                  "sitecore",
                  "content",
                  "Helix",
                ],
              },
            },
            "children": Array [
              BucketEntry {
                "bucket": Bucket {
                  "DB": "master",
                  "Filepath": "/data/helix/foundation/layer.project",
                  "Path": Path {
                    "Folder": "/sitecore/content/Helix",
                    "Name": "Project",
                    "Path": "/sitecore/content/Helix/Project",
                    "Tokens": Array [
                      "sitecore",
                      "content",
                      "Helix",
                      "Project",
                    ],
                  },
                },
                "children": Array [
                  BucketEntry {
                    "bucket": Bucket {
                      "DB": "master",
                      "Filepath": "/data/helix/project/test",
                      "Path": Path {
                        "Folder": "/sitecore/content/Helix/Project",
                        "Name": "TestProject",
                        "Path": "/sitecore/content/Helix/Project/TestProject",
                        "Tokens": Array [
                          "sitecore",
                          "content",
                          "Helix",
                          "Project",
                          "TestProject",
                        ],
                      },
                    },
                    "children": Array [],
                    "parent": [Circular],
                    "path": "/TestProject",
                  },
                ],
                "parent": [Circular],
                "path": "/Project",
              },
              BucketEntry {
                "bucket": Bucket {
                  "DB": "master",
                  "Filepath": "/data/helix/foundation/layer.foundation",
                  "Path": Path {
                    "Folder": "/sitecore/content/Helix",
                    "Name": "Foundation",
                    "Path": "/sitecore/content/Helix/Foundation",
                    "Tokens": Array [
                      "sitecore",
                      "content",
                      "Helix",
                      "Foundation",
                    ],
                  },
                },
                "children": Array [],
                "parent": [Circular],
                "path": "/Foundation",
              },
            ],
            "parent": [Circular],
            "path": "/Helix",
          },
        ],
        "parent": [Circular],
        "path": "/sitecore/content",
      },
      BucketEntry {
        "bucket": Bucket {
          "DB": "master",
          "Filepath": "/data/templates",
          "Path": Path {
            "Folder": "/sitecore",
            "Name": "templates",
            "Path": "/sitecore/templates",
            "Tokens": Array [
              "sitecore",
              "templates",
            ],
          },
        },
        "children": Array [],
        "parent": [Circular],
        "path": "/sitecore/templates",
      },
    ],
    "parent": null,
    "path": "",
  },
}
`;
